api/src/main/java/org/apache/iceberg/types/JavaHashes.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.types;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.IntFunction;
+import org.apache.iceberg.StructLike;
+
+public class JavaHashes {
+  private JavaHashes() {
+  }
+
+  public static int hashCode(CharSequence str) {
+    int result = 177;
+    for (int i = 0; i < str.length(); i += 1) {
+      char ch = str.charAt(i);
+      result = 31 * result + (int) ch;
+    }
+    return result;
+  }
+
+  static JavaHash<CharSequence> strings() {
+    return CharSequenceHash.INSTANCE;
+  }
+
+  static JavaHash<StructLike> struct(Types.StructType struct) {
+    return new StructLikeHash(struct);
+  }
+
+  static JavaHash<List<?>> list(Types.ListType list) {
+    return new ListHash(list);
+  }
+
+  private static class CharSequenceHash implements JavaHash<CharSequence> {
+    private static final CharSequenceHash INSTANCE = new CharSequenceHash();
+
+    private CharSequenceHash() {
+    }
+
+    @Override
+    public int hash(CharSequence str) {
+      if (str == null) {
+        return 0;
+      }
+
+      return JavaHashes.hashCode(str);
+    }
+  }
+
+  private static class StructLikeHash implements JavaHash<StructLike> {
+    private final JavaHash<Object>[] hashes;
+
+    private StructLikeHash(Types.StructType struct) {
+      this.hashes = struct.fields().stream()
+          .map(field ->
+            "unknown-partition".equals(field.doc()) ?
```

Comment:
Is `unknown-partition` some sort of a constant?

Reply:
Will it be safe to use `Objects::hashCode` in that case or shall we fail?

Reply:
This is an odd case. Without this, all tests are passing except for the forward-compatibility test that validates Spark can read tables with unknown transforms. When we see an unknown transform, we don't know what type of results it produces so we can't create a hash function or comparator ahead of time like we try to do here.

The actual problem is that when we create a hash function, it expects a `CharSequence` becuase our assumed result type is the most generic: string. That causes hashing to fail with a `ClassCastException`. The work-around, which is ugly, is to use the partition field doc to signal when this happens so that the hash function can use `Objects::hash` that can hash any object using its `hashCode` method, instead of trying to use our `CharSequence` hash function.

Reply:
I had an idea to fix this by making the hash function for strings check the value class. That's much cleaner. The method already has a null check so using `value instanceof CharSequence` isn't much more expensive. And, we should get a more consistent hash code by hashing `value.toString` than delegating to `Object#hashCode`.